sw/host/ot_certs/src/cwt.rs

```diff
@@ -0,0 +1,911 @@
+
+    for (name, var) in &template.variables {
+        let ret = vars.insert(name.clone(), CodegenVar::from_template_variable(name, var)?);
+        if ret.is_some() {
+            bail!("A variable name {} already exists", name);
+        }
+    }
+
+    for (name, var) in &template.constants {
+        let ret = vars.insert(name.clone(), CodegenVar::from_template_constant(name, var)?);
+        if ret.is_some() {
+            bail!("A constant name {} already exists", name);
+        }
+    }
+
+    Ok(vars)
+}
+
+fn generate_input_fields(vars: &CodegenVarTable, indent: &str) -> String {
+    vars.iter()
+        .filter(|(_, var)| !var.is_constant())
+        .map(|(_, var)| var.declarations(indent))
+        .collect()
+}
+
+fn generate_size_computations(size_exps: &[SizeExpression], prefix: &str) -> Result<String> {
+    let mut code = String::new();
+    for size_exp in size_exps.iter().filter(|exp| !exp.is_constant()) {
+        code += &format!(
+            "{prefix}const size_t content_size_{} = {};\n",
+            size_exp.index,
+            size_exp.content_expression()
+        );
+        code += &format!(
+            "{prefix}const size_t item_size_{} = {};\n",
+            size_exp.index,
+            size_exp.item_expression()
+        );
+    }
+
+    Ok(code)
+}
+
+fn collect_preorder_indicies_helper(nodes: &[CodegenStructure], root: usize, res: &mut Vec<usize>) {
+    res.push(root);
+    match &nodes[root] {
+        CodegenStructure::Var(_) => {}
+        CodegenStructure::CborByteArray(inner) => {
+            collect_preorder_indicies_helper(nodes, *inner, res);
+        }
+        CodegenStructure::Map(items) => {
+            for (k, v) in items {
+                collect_preorder_indicies_helper(nodes, *k, res);
+                collect_preorder_indicies_helper(nodes, *v, res);
+            }
+        }
+        CodegenStructure::Array(items) => {
+            for item in items {
+                collect_preorder_indicies_helper(nodes, *item, res);
+            }
+        }
+    }
+}
+
+fn collect_preorder_indicies(nodes: &[CodegenStructure]) -> Vec<usize> {
+    let mut res = Vec::new();
+    collect_preorder_indicies_helper(nodes, nodes.len() - 1, &mut res);
+    res
+}
+
+// Each CodegenStructure maps to a CodeBlock.
+type CodeBlock = Vec<GeneratedCode>;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum GeneratedCode {
+    // A single cbor-function call.
+    FunctionCall(String),
+    // If all the arguments are constants, derive its resulting bytes directly.
+    CborBytes(Vec<u8>),
+}
+
+struct GeneratedCborInstructions {
+    constant_definitions: String,
+    cbor_instructions: String,
+}
+
+fn generate_cbor_instructions(
+    nodes: &[CodegenStructure],
+    size_exps: &[SizeExpression],
+    prefix: &str,
+) -> Result<GeneratedCborInstructions> {
+    let call_wrapper = |func_call: String| {
+        indoc::formatdoc! { r#"
+        {prefix}HARDENED_RETURN_IF_ERROR({func_call});
```

Comment:
Check with @timothytrippel if hardening is really necessary and if so document why?

Reply:
As for the latest implementation, since we check the size before and after generating CBOR, we can make sure that we won't have any overflow error while in the middle of CBOR construction if no bugs in codegen. Therefore, we remove the error check for each individual CBOR function call now.

Reply:
@timothytrippel Do we need hardening here or we could just use the normal `RETURN_IF_ERROR`? I see X509 using the hardened one.

Reply:
I think RETURN_IF_ERROR is fine here. Where is the usage of the HARDENED version in X509 codegen?

Reply:
Oops, I remember the source wrong. I think I mix it with Tommy's PR.